pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -171,6 +211,140 @@ func referencesDiffs(old []metatypes.OwnerReference, new []metatypes.OwnerRefere
+// slice of a base type.
+func (gc *GarbageCollector) removeOrphanFinalizer(owner *node, retries int) error {
+	updated := false
+	for count := 0; count < retries; count++ {
+		ownerObject, err := gc.getObject(owner.identity)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot finalize owner %s, because cannot get it, tried %d times", owner.identity, count+1))
+			continue
+		}
+		accessor, err := meta.Accessor(ownerObject)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot access the owner object: %v, tried %d times", err, count+1))
+			continue
+		}
+		finalizers := accessor.GetFinalizers()
+		var newFinalizers []string
+		found := false
+		for _, f := range finalizers {
+			if f == api.OrphanFinalizer {
+				found = true
+			} else {
+				newFinalizers = append(newFinalizers, f)
+			}
+		}
+		if !found {
+			glog.V(6).Infof("the orphan finalizer is already removed from object %s", owner.identity)
+			updated = true
+			break
+		}
+		// remove the owner from dependent's OwnerReferences
+		ownerObject.SetFinalizers(newFinalizers)
+		_, err = gc.updateObject(owner.identity, ownerObject)
+		if err == nil {
+			updated = true
+			break
+		}
+		if err != nil && !errors.IsConflict(err) {
+			utilruntime.HandleError(fmt.Errorf("cannot update the finalizers of owner %s, with error: %v, tried %d times", owner.identity, err, count+1))
+			continue
+		}
+		// retry without reporting an error if it's a conflict
+		glog.V(6).Infof("got conflict updating the owner object %s, tried %d times", owner.identity, count+1)
+	}
+	if !updated {
+		return fmt.Errorf("updateMaxRetries(%d) has reached, giving up on finalizing the owner object %s. Please manually finalize it", retries, owner.identity)
+	}
+	return nil
+}
+
+// orphanFinalizer dequeues a node from the orphanQueue, then finds its dependents
+// based on the graph maintained by the GC, then removes it from the
+// OwnerReferences of its dependents, and finally updates the owner to remove
+// the "Orphan" finalizer. The node is add back into the orphanQueue if any of
+// these steps fail.
+func (gc *GarbageCollector) orphanFinalizer() {
+	key, quit := gc.orphanQueue.Get()
+	if quit {
+		return
+	}
+	defer gc.orphanQueue.Done(key)
+	owner, ok := key.(*node)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expect *node, got %#v", key))
+	}
+	// we don't need to lock each element, because they never get updated
+	owner.dependentsLock.RLock()
+	dependents := make([]*node, 0, len(owner.dependents))
+	for dependent := range owner.dependents {
+		dependents = append(dependents, dependent)
+	}
+	owner.dependentsLock.RUnlock()
+
+	// patch operation has less conflicts, so make its max retries lower
+	const patchMaxRetries = 3
+	err := gc.orhpanDependents(owner, dependents, patchMaxRetries)
+	if err != nil {
+		gc.orphanQueue.Add(owner)
+	}
+	// update the owner, remove "orphaningFinalizer" from its finalizers list
+	const updateMaxRetries = 5
```

Comment:
I removed the loop in patch, but kept the loop in update, because a update operation can easily get false conflicts.